[android][core] PlacePage for Tracks and Track Selection

kavikhalique opened the pull request:
This PR implements a basic display of place page for saved tracks. It shows the elevation graphs and also enables users to select and drag over the track as well as graph to see various values on track. 

It is similar to IOS implementation of  [#9785] 
Closes [#2829]
Closes [#2353]

It is in very initial stage for now I have just serialized all the present methods and classes to show the data. There are few known issues on which I will be working.

- [ ] Display correct co-ordinates of selected point on track
- [ ] Hide two same titles in place page and elevation graph

<img src="https://github.com/user-attachments/assets/c77a426c-1d8c-4e04-a7e7-881c393344b8" width="300">
<img src="https://github.com/user-attachments/assets/56d34459-5031-401f-823e-f8e1a85acfc6" width="300">


https://github.com/user-attachments/assets/a68c3f72-2adb-49b9-b899-4f84ecf5e718

biodranik commented:
I've launched an alpha build to test this PR: https://github.com/organicmaps/organicmaps/actions/runs/15155978936

kavikhalique commented on android/app/src/main/java/app/organicmaps/bookmarks/data/Track.java:
```diff
@@ -1,29 +1,56 @@
 
   Track(long trackId, long categoryId, String name, Distance length, int color)
   {
+    super(FeatureId.fromFeatureIdString("1:2:3"), BOOKMARK, name, "", "", "", 0, 0, "",
```
For normal tracks (not the one for place page) these data are not required so passing any value to constructor will not create any problem because these data will never be used. These data only get used when this track is used for place page and that time we are providing all the data. 

I was not sure what to provide in place of feature ID and Opening mode so I just chose a random value. Any better suggestion would be helpful.

biodranik commented:
@kavikhalique is it ready for public beta testing? What's missing?

biodranik commented:
Thanks a lot for your work, it works! Some feedback:

1. Track Place Page (TPP) should show the Edit button when a track is selected (and a graph for faster previewing). It is a general Android UX issue that it is not obvious to new users that PP can be expanded, and that an additional action is needed to edit a bookmark or track. A better alternative implementation is to show `[color icon   list name   edit pen icon]` like it is done in https://github.com/organicmaps/organicmaps/pull/10801 for iOS (then list name can be removed from the subtitle)
2. The share track button is missing near the X button
3. It is important to reduce the graph height so the expanded TPP size can also be reduced, to see more of the map (and the track). Current Ascent/Descent/Max elevation/Min elevation icons take a lot of space now, maybe reformatting them differently (e.g. like on iOS) is possible?
4. The graph is (slowly) redrawn when the phone is locked/unlocked. It looks weird. Is redrawing really needed?
5. Are unnecessary graph dots filtered out? Is one dot per pixel drawn now? It may speed up the rendering
6. @kirylkaveryn @euf is (improved) edit button better at the bottom or at the top of TPP? At the bottom, it can be pressed more easily, but at the top (above the graph) it is more consistent with bookmarks PP.
7. TPP changes its height if device is locked/unlocked. PP works fine in such a case.

kavikhalique commented:
Thanks for such a detailed feedback. I will apply the changes as suggested in the above review.

kirylkaveryn commented:
> @kirylkaveryn @euf is (improved) edit button better at the bottom or at the top of TPP? At the bottom, it can be pressed more easily, but at the top (above the graph) it is more consistent with bookmarks PP.


At the bottom of the place page the Edit button will be harder to discover and always require additional number taps or swipes  (which we prefer to decrease as much as possible) to reach.

Since the changing colour and group may be relatively frequently used it should be always visible and easy accessible.

Also, if we implement easy title editing it will be strange and not very good UI to spread the editing  of single entity actions across the whole screen.

biodranik commented:
> At the bottom of the place page the Edit button will be harder to discover and always require additional number taps or swipes (which we prefer to decrease as much as possible) to reach.
> 
> Since the changing colour and group may be relatively frequently used it should be always visible and easy accessible.
> 
> Also, if we implement easy title editing it will be strange and not very good UI to spread the editing of single entity actions across the whole screen.

All are valid points, thanks! @kavikhalique so it's better to move the Edit cell above the graph.

The elevation ▲ icon doesn't look smooth on the graph. Is it possible to draw emoji ▲ there?

biodranik commented:
Crash when selecting a track from Bookmarks and Tracks (if rebase on the latest master)

biodranik commented:
@kavikhalique here is the fix: https://github.com/organicmaps/organicmaps/commit/00270cf144235b0a7e8e19e0fe8fe906592a688f

Some tracks were clickable, while some were crashing.

biodranik commented:
- `Dist:` text on the selected point on the graph looks weird on a separate line, maybe replace it with unicode/emoji like ↦ or ⇥ or → or ↔ ?
- When my current position is at the center, I select a track from bookmarks, the app zooms to it. When I close TPP, map zooms back to my position instead of staying on the track.

kavikhalique commented:
<img src="https://github.com/user-attachments/assets/17273865-b131-4e33-a804-6b897799e5fa" width="250">
<img src="https://github.com/user-attachments/assets/b474454d-dda7-4122-8fd2-1c444a62a046" width="250">
<img src="https://github.com/user-attachments/assets/ed01a4b6-d2f9-4758-85b0-5ff5cc92f8bc" width="250">
<img src="https://github.com/user-attachments/assets/e3dd220f-2129-4301-8dc1-f85197ca53a3" width="250">
<img src="https://github.com/user-attachments/assets/b57d49ee-1104-4e32-8cab-78b4efd096df" width="250">

biodranik commented:
Please integrate the crash fix commit into your branch https://github.com/organicmaps/organicmaps/commit/5fee3db0ad0ecd3c39407c9ade7a7ecca46caa32

biodranik commented:
Thanks! Looks great! A few more improvements:

1. Remove the duplicated list name from the TPP subtitle
2. Remove "Time: x min" at the bottom of the graph
3. Graph should be immediately visible if it is available, without additional action to expand TPP
4. There are some glitches in how the first and the last points on the graph are rendered.
5. Share track button should show popup menu with 2 choices like in Bookmarks and Tracks (Export GPX and Export KMZ) (now it shares some geo: point)
6. Please check that "save bookmarks into the last used list" logic is working properly.

![telegram-cloud-photo-size-2-5463090416900502076-y](https://github.com/user-attachments/assets/dbd968dc-f976-4936-852c-416c6531a51c)
![telegram-cloud-photo-size-2-5463090416900502077-y](https://github.com/user-attachments/assets/bbce846a-d08e-4a7d-a2c8-6fc52682c343)

biodranik commented:
> I am not sure what exactly it means. Does it means slow drawing of graph?

Take a look at screenshots:
1. The last point shows different elevation (659m instead of 660m) and the popup with text is behind the graph's label 660m (should be on front of it).
2. The first point's circle is cut off at the top.

Is GPX/KMZ sharing fixed too?

kavikhalique commented:
> > I am not sure what exactly it means. Does it means slow drawing of graph?
> 
> 
> 
> Take a look at screenshots:
> 
> 1. The last point shows different elevation (659m instead of 660m) and the popup with text is behind the graph's label 660m (should be on front of it).
> 
> 2. The first point's circle is cut off at the top.
> 
> 

Oh thanks for clearance i think the min and max altitude value on graph are shown incorrectly will fix it.

> Is GPX/KMZ sharing fixed too?

Probably need to implement a separate method in jni for single track export. Currently for exporting a single track in android it only shows links instead of file export. (Can be cross checked by exporting a single track from track lists.)

biodranik reviewed (commented):
Thanks! Already better 👍 A few more fixes are needed:
- [x] Remove Edit Bookmark button
- [ ] Remove duplicated list name from the PP title (there was a todo left by @kirylkaveryn for Android)
- [ ] Remove "Category" title for edit cell, it is misleading and takes important vertical space. We'll think about better hints for new users later.
- [ ] Multiline list name looks wrong, is it possible to increase the height of the cell so longer name can fit? E.g. limit it by 3 lines and then use … at the end?
- [x] GPX/KMZ sharing

<img width="1440" height="3120" alt="grafik" src="https://github.com/user-attachments/assets/bcf7cd71-623d-487e-bc04-560e44584aa6" />

kavikhalique commented:
Thanks, will be pushing changes by tonight.

> - [x] GPX/KMZ sharing

This one is already implemented, kindly test it.

biodranik commented:
> This one is already implemented, kindly test it.

I tried to rebase this MR on the latest master and test, and share button only shared some geo: link, as before…

kavikhalique commented:
Updated and Rebased the PR.

<img width="300" alt="Screenshot_20250721-012135" src="https://github.com/user-attachments/assets/a9e0a515-41c9-4fda-8f6c-017bb03a99ad" />
<img width="300" alt="Screenshot_20250721-012215" src="https://github.com/user-attachments/assets/d714c3b2-260f-47cf-be8e-42533518c6a5" />

biodranik commented:
@kavikhalique please rebase on the latest master and fix these errors:
```
/Users/alex/Developer/omim/omim/android/app/src/main/java/app/organicmaps/sdk/bookmarks/data/Bookmark.java:116: error: incompatible types: int cannot be converted to Parcel
    Icon icon = new Icon(PredefinedColors.getPredefinedColorIndex(color));
                                                                 ^
/Users/alex/Developer/omim/omim/android/app/src/main/java/app/organicmaps/widget/placepage/PlacePageView.java:531: error: cannot find symbol
        args.putInt(BookmarkColorDialogFragment.ICON_TYPE, PredefinedColors.getPredefinedColorIndex(mTrack.getColor()));
                                               ^
  symbol:   variable ICON_TYPE
  location: class BookmarkColorDialogFragment
/Users/alex/Developer/omim/omim/android/app/src/main/java/app/organicmaps/widget/placepage/PlacePageView.java:546: error: cannot find symbol
        args.putInt(BookmarkColorDialogFragment.ICON_TYPE, bookmark.getIcon().getColor());
                                               ^
  symbol:   variable ICON_TYPE
  location: class BookmarkColorDialogFragment
```

@AndrewShkrob can you help please?

kavikhalique commented:
> @kavikhalique please rebase on the latest master and fix these errors:

Done

kavikhalique commented:
- [x] Delete Track button (can add it together with Save button now on the bottom bar, iOS doesn't have Save button, we need to think about a better UX there. Maybe put it near the pen icon?)

Added 4th bottom sheet button as delete button, this button will not be available during routing mode. Also added one alert dialog to prevent accidental tap deletion of tracks.

- [x] Is it possible to layout statistics icons and their texts similar to iOS (above the graph, and in a similar layout, to have larger fonts)?

Matched the UI of iOS and decreased the overall size of place page.

- [ ] Graph is rendered noticeably slowly, can its implementation be faster?

May be it can be improved later in further PRs

Attached previews:

<img width="300" alt="1000000077" src="https://github.com/user-attachments/assets/c876736c-df09-4224-afd4-a4db950fd903" />
<img width="300" alt="1000000078" src="https://github.com/user-attachments/assets/d3eadb05-343f-4494-850e-96e29d582d8b" />

biodranik commented:
@kavikhalique thanks! Looks way better now 🔥 

1. Is it possible to align "Ascent" and other texts with icons on top of them, like on iOS?
2. Can you please rebase and check/fix any conflicts? Likely, you always need to choose your own implementation and then apply formatting. I suggest squashing all your commits first before doing the rebase, and then you can extract independent/logical ones again. 
3. As we discussed with @kirylkaveryn and @euf, the delete track button is related to the track, and route from/to and save buttons are related to a point, so it is a bit confusing, so we'll think about a better UI/UX (but for beta testing it is already acceptable).

kavikhalique commented:
Sure will do it 👍

biodranik commented:
@kavikhalique we've merged clang-format code style checks right after you rebased it :) Can you please rebase your code again and apply formatting with clang-format? There is a script tools/unix/clang-format.sh if you didn't set it up using Android Studio plugin https://plugins.jetbrains.com/plugin/20785-clang-format-tools

biodranik commented:
Failing debug iOS test is a known issue https://github.com/organicmaps/organicmaps/issues/9867
I suspect that tests actually are running and are failing there sometimes, but we don't see the full log output. @kirylkaveryn is it possible to see all OM logs when running tests in iOS sim?

biodranik commented:
@kavikhalique thanks! Can you please remove the "Save..." dialog when route recording is stopped, and show the saved track place page immediately instead?

We also need to think about better UX for bottom bar buttons, esp. Delete, which is related to the track, compared to Route From, Save, Route To, which are related to the point on the map.

biodranik commented:
- [ ] There is still a bug with overlapping texts in the bottom right corner, the selected point's label should be above.
- [ ] The top left max elevation value is not rendered on the graph

<img width="945" height="2048" alt="выява" src="https://github.com/user-attachments/assets/241549c7-b38d-427a-b013-a83b8187ef05" />

vng commented on android/app/src/main/java/app/organicmaps/sdk/bookmarks/data/Track.java:
```diff
@@ -52,4 +91,25 @@ public String getTrackDescription()
+  {
+    if (mElevationInfo != null)
+      return mElevationInfo;
+    mElevationInfo = BookmarkManager.nativeGetTrackElevationInfo(mTrackId);
```
nit:
```
if (x == null)
  x = ...
return x;
```

vng commented on android/app/src/main/java/app/organicmaps/sdk/bookmarks/data/Track.java:
```diff
@@ -52,4 +91,25 @@ public String getTrackDescription()
+    if (mTrackStatistics != null)
+      return mTrackStatistics;
+    mTrackStatistics = BookmarkManager.nativeGetTrackStatistics(mTrackId);
+    return mTrackStatistics;
```
nit:
```
if (x == null)
  x = ...
return x;
```

vng commented on android/sdk/src/main/cpp/app/organicmaps/sdk/bookmarks/data/BookmarkManager.cpp:
```diff
@@ -858,4 +866,41 @@ Java_app_organicmaps_sdk_widget_placepage_PlacePageButtonFactory_nativeHasRecent
+{
+  return static_cast<jboolean>(frm()->GetBookmarkManager().GetTrack(track_id)->GetElevationInfo().has_value());
+}
+JNIEXPORT jobject JNICALL Java_app_organicmaps_sdk_bookmarks_data_BookmarkManager_nativeGetTrackStatistics(
```
clang-format ?

biodranik replied:
```suggestion

JNIEXPORT jobject JNICALL Java_app_organicmaps_sdk_bookmarks_data_BookmarkManager_nativeGetTrackStatistics(
```

vng commented on android/sdk/src/main/cpp/app/organicmaps/sdk/bookmarks/data/BookmarkManager.cpp:
```diff
@@ -858,4 +866,41 @@ Java_app_organicmaps_sdk_widget_placepage_PlacePageButtonFactory_nativeHasRecent
+  auto const & elevationInfo = track->GetElevationInfo();
+  return usermark_helper::CreateElevationInfo(env, elevationInfo.value());
+}
+JNIEXPORT jboolean JNICALL Java_app_organicmaps_sdk_bookmarks_data_BookmarkManager_nativeIsElevationInfoHasValue(
```
nit: ```return null;``` in nativeGetTrackElevationInfo instead of having a second JNI function ```hasXXX```?

biodranik replied:
```suggestion

JNIEXPORT jboolean JNICALL Java_app_organicmaps_sdk_bookmarks_data_BookmarkManager_nativeIsElevationInfoHasValue(
```

vng reviewed (commented):
I propose to merge this PR and move forward with fixes.

biodranik commented on android/app/src/main/java/app/organicmaps/widget/placepage/PlacePageViewModel.java:
```diff
@@ -33,6 +37,38 @@ public void setMapObject(MapObject mapObject)
     mMapObject.setValue(mapObject);
   }
 
+  // These silent method are used to update data of map object silently without triggering
+  // the observer which refreshes everything and puts extra load  on device in calculation
+  // which is not required
+  public void modifyMapObjectPointSilently(ElevationInfo.Point point)
+  {
+    if (mMapObject.getValue() == null)
+      return;
+    mMapObject.getValue().setLat(point.getLatitude());
+    mMapObject.getValue().setLon(point.getLongitude());
+  }
+
+  public void modifyMapObjectCategoryIdSilently(long categoryId)
+  {
+    if (mMapObject.getValue() == null)
+      return;
+    switch (PlacePageView.MapObjectType.getMapObjectType(mMapObject.getValue()))
+    {
+    case TRACK -> ((Track) mMapObject.getValue()).setCategoryId(categoryId);
+    case BOOKMARK -> ((Bookmark) mMapObject.getValue()).setCategoryId(categoryId);
+    }
+  }
+
+  public void modifyMapObjectColorSilently(int color)
+  {
+    if (mMapObject.getValue() == null)
+      return;
+    switch (PlacePageView.MapObjectType.getMapObjectType(mMapObject.getValue()))
+    {
+    case TRACK -> ((Track) mMapObject.getValue()).setColor(color);
+    case BOOKMARK -> ((Bookmark) mMapObject.getValue()).setIconColor(color);
+    }
+  }
```
Isn't it an unnecessary complication? There are already `mMapObject` objects casted to the necessary Track or Bookmark objects where you call these functions. Calling there `mTrack.setColor(color)` or other functions would make the code cleaner, no?

```suggestion
```

kavikhalique replied:
I have used it in many places other than that also mostly for avoiding if else and using switch cases. In future if any other type increases readability will decrease with if else WDYT

biodranik commented on android/app/src/main/java/app/organicmaps/widget/placepage/PlacePageView.java:
```diff
@@ -796,4 +1067,21 @@ public interface PlacePageViewListener
     void onPlacePageRequestToggleState();
     void onPlacePageRequestClose();
   }
+
+  public enum MapObjectType
+  {
+    TRACK,
+    BOOKMARK,
+    OTHER;
+
+    public static MapObjectType getMapObjectType(MapObject mapObject)
+    {
+      if (mapObject.isTrack())
+        return TRACK;
+      else if (mapObject.isBookmark())
+        return BOOKMARK;
+      else
+        return OTHER;
+    }
+  }
```
This aux enum is also not needed if MapObject is modified directly, without the code in PlacePageViewModel, right?
```suggestion
```

kavikhalique replied:
I have used it in many places other than that also mostly for avoiding if else and using switch cases. In future if any other type increases readability will decrease with if else WDYT

biodranik commented on android/app/src/main/java/app/organicmaps/widget/placepage/sections/PlacePageTrackFragment.java:
```diff
@@ -0,0 +1,115 @@
+      return;
+    mElevationProfileViewRenderer.onChartElevationActivePointChanged();
+    ElevationInfo.Point point = BookmarkManager.INSTANCE.getElevationActivePointCoordinates(mTrack.getTrackId());
+    mViewModel.modifyMapObjectPointSilently(point);
```
Would it work?
```suggestion
    mTrack.setLat(point.getLatitude());
    mTrack.setLon(point.getLongitude());
```

biodranik commented on android/app/src/main/java/app/organicmaps/widget/placepage/PlacePageView.java:
```diff
@@ -420,6 +464,180 @@ private void refreshPreview()
+        return;
+      BookmarkManager.INSTANCE.notifyCategoryChanging(track, newCategory.getId());
+      mTvCategory.setText(newCategory.getName());
+      mViewModel.modifyMapObjectCategoryIdSilently(newCategory.getId());
```
Would it work?
```suggestion
      track.setCategory(newCategory.getId());
```

biodranik commented on android/app/src/main/java/app/organicmaps/widget/placepage/PlacePageView.java:
```diff
@@ -420,6 +464,180 @@ private void refreshPreview()
+      if (previousCategory == newCategory)
+        return;
+      mTvCategory.setText(newCategory.getName());
+      mViewModel.modifyMapObjectCategoryIdSilently(newCategory.getId());
```
Would it work?
```suggestion
      bookmark.setCategory(newCategory.getId());
```

biodranik commented on android/app/src/main/java/app/organicmaps/widget/placepage/PlacePageView.java:
```diff
@@ -420,6 +464,180 @@ private void refreshPreview()
+        int to = PredefinedColors.getColor(colorPos);
+        if (from == to)
+          return;
+        mViewModel.modifyMapObjectColorSilently(to);
```
Would it work?
```suggestion
        bookmark.setIconColor(to);
```

biodranik commented on android/app/src/main/java/app/organicmaps/widget/placepage/PlacePageView.java:
```diff
@@ -420,6 +464,180 @@ private void refreshPreview()
+        int to = PredefinedColors.getColor(colorPos);
+        if (from == to)
+          return;
+        mViewModel.modifyMapObjectColorSilently(to);
```
Would it work?
```suggestion
        track.setColor(to);
```

biodranik commented on android/app/src/main/java/app/organicmaps/widget/placepage/PlacePageView.java:
```diff
@@ -420,6 +464,180 @@ private void refreshPreview()
+    {
+    case TRACK ->
+    {
+      final Track mTrack = (Track) mMapObject;
```
```suggestion
      final Track track = (Track) mMapObject;
```

biodranik commented on android/app/src/main/java/app/organicmaps/widget/placepage/sections/PlacePageTrackFragment.java:
```diff
@@ -0,0 +1,115 @@
+    // MapObject could be something else than a Track if the user already has the place page
+    // opened and clicks on a non-Track POI.
+    // This callback would be called before the fragment had time to be destroyed
+    if (mapObject != null && mapObject.isTrack())
```
nit to reduce indentation:
```suggestion
    if (mapObject == null || !mapObject.isTrack())
      return;
```

biodranik commented on android/app/src/main/res/drawable/ic_distance_travelled.xml:
Where is this icon from?

kavikhalique replied:
I downloaded it from svgrepo : )

biodranik replied:
Can it be replaced with unicode character ↔ or corresponding google material icon https://fonts.google.com/icons?selected=Material+Symbols+Outlined:arrow_range:FILL@0;wght@400;GRAD@0;opsz@24&icon.query=arrow&icon.size=24&icon.color=%23e3e3e3 , or any other source already listed in data/copyright.html ?

biodranik commented on android/app/src/main/res/layout/elevation_profile_internal.xml:
```diff
@@ -4,165 +4,202 @@
-    </LinearLayout>
-    <LinearLayout
+    </RelativeLayout>
+    <!--card four-->
```
What are these cards for?

kavikhalique replied:
These are 4 cards containing ascent descent min elevation and max elevation data on pp

biodranik commented on android/app/src/main/res/values/strings.xml:
```diff
@@ -924,4 +924,5 @@
     <string name="uri_open_location_failed">No app installed that can open the location</string>
     <!-- preference string for using auto theme only in navigation mode -->
     <string name="nav_auto">Auto in navigation</string>
+    <string name="PP_track_bottom_sheet_title">Save Track As</string>
```
```suggestion
    <!-- Title of the bottom sheet when sharing the track -->
    <string name="share_track">Share Track</string>
```

biodranik commented on android/sdk/src/main/cpp/app/organicmaps/sdk/UserMarkHelper.cpp:
```diff
@@ -88,6 +88,47 @@ jobject CreateMapObject(JNIEnv * env, place_page::Info const & info, int mapObje
+
+  auto const trackId = info.GetTrackId();
+  auto const track = frm()->GetBookmarkManager().GetTrack(trackId);
+  dp::Color nColor = track->GetColor(0);
+
+  jint androidColor =
+      shift(nColor.GetAlpha(), 24) + shift(nColor.GetRed(), 16) + shift(nColor.GetGreen(), 8) + nColor.GetBlue();
```
```suggestion
  jint const androidColor = track->GetColor(0).GetARGB();
```

biodranik commented on android/sdk/src/main/cpp/app/organicmaps/sdk/bookmarks/data/BookmarkManager.cpp:
```diff
@@ -447,6 +447,16 @@ JNIEXPORT jobject JNICALL Java_app_organicmaps_sdk_bookmarks_data_BookmarkManage
   return usermark_helper::CreateMapObject(env, g_framework->GetPlacePageInfo());
 }
 
+JNIEXPORT void JNICALL Java_app_organicmaps_sdk_bookmarks_data_BookmarkManager_nativeUpdateTrackPlacePage(JNIEnv * env,
+                                                                                                          jobject,
+                                                                                                          jlong trackId)
```
nit:, also, should trackId parameter be removed?:
```suggestion
JNIEXPORT void JNICALL Java_app_organicmaps_sdk_bookmarks_data_BookmarkManager_nativeUpdateTrackPlacePage(JNIEnv *,                                                                                                          jobject,                                                                                                     jlong)
```

biodranik commented on android/sdk/src/main/cpp/app/organicmaps/sdk/bookmarks/data/BookmarkManager.cpp:
```diff
@@ -824,14 +834,12 @@ Java_app_organicmaps_sdk_bookmarks_data_BookmarkManager_nativeRemoveElevationCur
-       0},  // todo(KK): replace with coordinates from the elevation profile point to show selection mark on the track
-      static_cast<double>(distanceInMeters));
+  bm.SetElevationActivePoint(static_cast<kml::TrackId>(trackId),
+                             m2::PointD(static_cast<double>(latitude), static_cast<double>(longitude)),
+                             static_cast<double>(distanceInMeters));
```
Is static cast really needed?
```suggestion
                             {latitude, longitude}, distanceInMeters);
```

biodranik commented on android/sdk/src/main/cpp/app/organicmaps/sdk/bookmarks/data/BookmarkManager.cpp:
```diff
@@ -858,4 +866,41 @@ Java_app_organicmaps_sdk_widget_placepage_PlacePageButtonFactory_nativeHasRecent
+  static jmethodID const cId = jni::GetConstructorID(env, g_trackStatisticsClazz, "(DDDDII)V");
+  auto const trackStats = frm()->GetBookmarkManager().GetTrack(track_id)->GetStatistics();
+  return env->NewObject(g_trackStatisticsClazz, cId, static_cast<jdouble>(trackStats.m_length),
+                        static_cast<jdouble>(trackStats.m_duration), static_cast<jdouble>(trackStats.m_ascent),
+                        static_cast<jdouble>(trackStats.m_descent), static_cast<jint>(trackStats.m_minElevation),
```
Is casting really needed?
```suggestion
                        trackStats.m_duration, trackStats.m_ascent,
                        trackStats.m_descent, static_cast<jint>(trackStats.m_minElevation),
```

In `android-sdk/ndk/28.2.13676358/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr/include/jni.h`:
```
typedef double   jdouble;  /* 64-bit IEEE 754 */
```

biodranik reviewed (commented):
Thanks a lot, we're almost there! I mentioned several important fixes in a comment above, mainly these two:

- Remove Save recorded track dialog confirmation and immediately show saved track's place page instead
- For some reason, "Add place to OpenStreetMap" button appears when any building is selected (see screenshot below). It shouldn't be visible if Add Business is already displayed.

You can review comments, fix most critical ones, so it would be safe to merge into master, and we can continue improvements in iterations.

biodranik commented:
> It will directly stop without asking.

Agree, let's do it properly as on iOS in the next iteration:
- When the track is recorded, pressing recording button shows RTPP (Recording Track Place Page), which may be empty, with a Save button on the bottom toolbar. The recording button at the top right is hidden.
- Pressing Save button saves the track and shows TPP (Track Place Page).
- Closing RTPP shows recording button at the top right again.

biodranik commented:
Edit Track or Bookmark pencil icon is hardly tappable, can its tap area be increased? Maybe the cell height and icon itself can be slightly bigger?

Another improvement: add a Delete Track button into its editing dialog.

kavikhalique commented on android/app/src/main/java/app/organicmaps/widget/placepage/PlacePageController.java:
```diff
@@ -202,18 +217,6 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
     ViewCompat.requestApplyInsets(mPlacePage);
   }
 
-  @NonNull
-  private static PlacePageButtons.ButtonType toPlacePageButton(@NonNull RoadWarningMarkType type)
-  {
-    return switch (type)
-    {
-      case DIRTY -> PlacePageButtons.ButtonType.ROUTE_AVOID_UNPAVED;
-      case FERRY -> PlacePageButtons.ButtonType.ROUTE_AVOID_FERRY;
-      case TOLL -> PlacePageButtons.ButtonType.ROUTE_AVOID_TOLL;
-      default -> throw new AssertionError("Unsupported road warning type: " + type);
-    };
-  }
-
```
This happened during rebase conflict resolution. There is no use of this code in place page.

kavikhalique replied:
done committing soon..Since this PR Increased so much it became hard to resolve lots of conflicts and in between I missed it sorry for inconvenience : |

biodranik replied:
No problem, thanks for your hard work, Kavi!

kavikhalique commented on android/app/src/main/java/app/organicmaps/widget/placepage/PlacePageController.java:
```diff
@@ -439,6 +443,48 @@ private void onBookmarkBtnClicked()
+      return;
+    if (mAlertDialog != null)
+    {
+      mAlertDialog.dismiss();
+      mAlertDialog.show();
```
In a hypothetical condition when anyhow delete button get pressed twice it may trigger multiple dialogs so it basically disables previous dialogs

biodranik replied:
I see. MWMActivity is a different beast. Here, however, there is a different issue. mAlertDialog should be recreated for each track to include its title (KUDOS for the good UX idea btw!).

Is it possible to nullify mAlertDialog on user's action? If yes, then if (mAlertDialog != null) will mean that dialog is opened and should be dismissed first.

biodranik commented on android/app/src/main/java/app/organicmaps/widget/placepage/PlacePageView.java:
```diff
@@ -327,6 +361,10 @@ private void refreshViews()
       refreshMyPosition(loc);
     else
       refreshDistanceToObject(loc);
+    UiUtils.hideIf(mMapObject.isTrack(), mFrame.findViewById(R.id.ll__place_latlon),
+                   mFrame.findViewById(R.id.ll__place_open_in), mFrame.findViewById(R.id.ll__place_add), mEditTopSpace,
+                   mTvAzimuth, mTvDistance, mAvDirection);
+    UiUtils.hideIf(MapObjectType.getMapObjectType(mMapObject) != MapObjectType.OTHER, mTvSubtitle);
```
The first line hides fields for track. This second line hides mTvSubtitle for both bookmarks and tracks, why?

kavikhalique replied:
In bookmark's subtitle we have the category name so it repeats similar to tracks. Should i keep it for bookmarks?

biodranik replied:
I see, just adding a comment about it would be enough.

kavikhalique commented on android/app/src/main/java/app/organicmaps/widget/placepage/PlacePageController.java:
```diff
@@ -439,6 +443,48 @@ private void onBookmarkBtnClicked()
 
+  private void onTrackRemoveClicked()
+  {
+    // mMapObject is set to null when the place page closes
+    // We don't want users to interact with the buttons when the PP is closing
+    if (mMapObject == null)
+      return;
```
This comment I copied from already present methods they have used the same check so I also added that.

biodranik replied:
Yes, that's why I'm asking if there's a cleaner way to disable user interaction instead of copy-pasting the same check in every function.

kavikhalique replied:
The problem which i see in disabling the bottom sheet buttons is that there are many other clickable things on pp which can be clicked while pp is closing so may be disabling only bottomsheet will not solve the problem from root. We should instead disable the touch interaction of whole PP to avoid any interactions during pp close.

biodranik replied:
Disabling touch interaction properly is the key. Maybe a separate issue can be created, and comments from code can reference it, so we'll investigate it later. I wonder if there are other, yet uncovered bugs in other places where these explicit checks are not present.

vng reviewed (approved):
Did some tests, LGTM, ready to merge.

kavikhalique commented:
> Thanks, Kavi! Is there anything left to fix from this PR's comments?
> 

Removal of dialog box while turning off Track Recording and showing PP. I am currently working on it. Will raise draft PR soon. If there is anything else kindly inform me.

> Looks like there's also a crash: #11063

it has already been fixed by followup PRs : )
